### ISSUE TITLE
Dataset.unbatch() sets cardinality to -2 when batch remainder is not explicitly dropped

### ISSUE THREAD

**Issue by christian-steinmeyer** [tensorflow/tensorflow#44933]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): windows 10
- TensorFlow installed from (source or binary): pip
- TensorFlow version (use command below): 2.3.1
- Python version: 3.8.5

**Describe the current behavior**
Using `tf.data.experimental.cardinality()` yields `-2` (unknown), when batching (without explicitly dropping the remainder) and unbatching a dataset.

**Describe the expected behavior**
`tf.data.experimental.cardinality()` yields the same result as `len(list(tf.data.Dataset.as_numpy_iterator()))`.

**Standalone code to reproduce the issue**

```python
    import tensorflow as tf

    print(tf.version.VERSION)  # 2.3.1
    ds = tf.data.Dataset.range(10)  # shape=()
    ds = ds.batch(2)  # shape=(2,)
    print(tf.data.experimental.cardinality(ds))  # 5
    ds = ds.unbatch()  # shape=()
    print(len(list(ds.as_numpy_iterator()))) # 10
    print(tf.data.experimental.cardinality(ds))  # Should be 10, but is -2 (unknown)
```

**Other info / logs**
The issue does not occur (anymore), when using `drop_remainder=True` (see #39136)

**Comment by bhack** [comment]
See the difference of the dataset `element_spec` with `True and False`:
```
ds = ds.batch(2,drop_remainder=False)
print("Element spec:",ds.element_spec)
```
And check the the `lambda` in the two cases:
https://github.com/tensorflow/tensorflow/blob/df86a9308a1112a8fd6cd8369a24ec84fa6cd125/tensorflow/python/data/ops/dataset_ops.py#L3906-L3917

So in one case when we are finalizing the `unbatch` operation we have no more the `element_spec`

https://github.com/tensorflow/tensorflow/blob/df86a9308a1112a8fd6cd8369a24ec84fa6cd125/tensorflow/python/data/ops/dataset_ops.py#L4665-L4667

**Comment by Saduf2019** [comment]
I am able to replicate this on 2.3 and nightly, please find the [gist here](https://colab.research.google.com/gist/Saduf2019/74e5d6d0f3578928327db8241740e003/untitled463.ipynb).

**Comment by aaudiber** [comment]
The problem is that the shape passed to `unbatch` is unknown. If `unbatch` can't determine the shape of its inputs, it can't determine its cardinality.

```
    ds = tf.data.Dataset.range(10)  # shape=()
    ds = ds.batch(2)
    print(ds.element_spec.shape)  # prints (None,)
```

Shape inference is not currently smart enough to realize that because 2 evenly divides 10, the shape will always be (2,). Instead, shape inference is conservative and says the shape is unknown, thinking there may be a partial batch at the end. Fixing this is not straightforward, because cardinalities are not known during shape inference. In particular, when computing the shape of the elements produced by `ds.batch(2)`, all we know is that the input shapes are scalars and that `drop_remainder` is `False`.

It is possible to make cardinality information available to shape inference, but for it to work across both eager and graph modes we would need to add Python implementations of cardinality for all types of datasets. Instead of duplicating cardinality logic, we would probably want to make the Python cardinality implementations the source of truth.

In summary, I think this is a valid feature request that requires some significant refactoring to achieve. In the short term, it is expected that we can't infer the shape of a batch dataset unless drop_remainder is true.

**Comment by bhack** [comment]
@aaudiber:

In the mean time what do you think about a small PR like (as it will be cover some eager cases)?

```
    try:
      modulo = self._input_dataset.__len__() % self._batch_size
    except TypeError:
      modulo = True
    # pylint: disable=protected-access
    if constant_drop_remainder or (not modulo) :

```

**Comment by aaudiber** [comment]
@bhack I'd prefer to keep the behavior consistent between eager and graph mode, so that putting a @tf.function around eager code won't change the result of cardinality/__len__.

**Comment by bhack** [comment]
> @bhack I'd prefer to keep the behavior consistent between eager and graph mode, so that putting a @tf.function around eager code won't change the result of cardinality/**len**.

OK but `len` is already conditional:

https://github.com/tensorflow/tensorflow/blob/df86a9308a1112a8fd6cd8369a24ec84fa6cd125/tensorflow/python/data/ops/dataset_ops.py#L447-L449

**Comment by aaudiber** [comment]
Ah I see, in that case I think your suggestion makes sense for `len`.

**Comment by bhack** [comment]
> Ah I see, in that case I think your suggestion makes sense for `len`.

Ok. Can you check https://github.com/tensorflow/tensorflow/pull/45047

**Comment by jamesbraza** [comment]
For what it's worth, I am hitting this right now as well.  What's too bad is I am using [`image_dataset_from_directory`](https://www.tensorflow.org/api_docs/python/tf/keras/utils/image_dataset_from_directory) so I don't even have the option to set a `drop_remainder=True`.  My version is `tensorflow==2.4.4`.

**Comment by Venkat6871** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.
